src/persist-client/src/read.rs

```diff
@@ -553,129 +508,116 @@ where
-    /// exchanged (including over the network) to load balance the processing of
-    /// this snapshot. These splits are usable by anyone with access to the
-    /// shard's [crate::PersistLocation]. The `len()` of the returned `Vec` is
-    /// `num_splits`. If a 1:1 mapping between splits and (e.g. dataflow
-    /// workers) is used, then the work of replaying the snapshot will be
-    /// roughly balanced.
-    ///
-    /// This method exists to allow users to parallelize snapshot iteration. If
-    /// you want to immediately consume the snapshot from a single place, you
-    /// likely want the [Self::snapshot] helper.
+    /// For more details on this operation's semantics, see [Self::snapshot] and
+    /// [Self::listen].
     #[instrument(level = "debug", skip_all, fields(shard = %self.machine.shard_id()))]
-    pub async fn snapshot_splits(
-        &self,
-        as_of: Antichain<T>,
-        num_splits: NonZeroUsize,
-    ) -> Result<Vec<SnapshotSplit<T>>, Since<T>> {
-        trace!(
-            "ReadHandle::snapshot as_of={:?} num_splits={:?}",
-            as_of,
-            num_splits
-        );
-        // Hack: Keep this method `&self` instead of `&mut self` by cloning the
-        // cached copy of the state, updating it, and throwing it away
-        // afterward.
-        let mut machine = self.machine.clone();
-        let batches = machine.snapshot(&as_of).await?;
-        let batches = batches.into_iter().flat_map(|b| {
-            let desc = b.desc.clone();
-            b.keys.into_iter().map(move |k| (k, desc.clone()))
-        });
-        let mut splits = Vec::new();
-        for _ in 0..num_splits.get() {
-            // Register this split as a new reader so that it gets its own
-            // seqno capability.
-            let split_reader_id = ReaderId::new();
-            let _read_cap = machine
-                .clone_reader(&split_reader_id, (self.cfg.now)())
-                .await;
-            // We don't want the frontier capability, so explictly toss it.
-            let _ = machine
-                .downgrade_since(&split_reader_id, &Antichain::new(), (self.cfg.now)())
-                .await;
-            splits.push(SnapshotSplit {
-                reader_id: split_reader_id,
-                shard_id: self.machine.shard_id(),
-                as_of: as_of.clone(),
-                batches: Vec::new(),
-            })
-        }
-        for (idx, (batch_key, desc)) in batches.into_iter().enumerate() {
-            splits[idx % num_splits.get()]
-                .batches
-                .push((batch_key, desc.clone()));
-        }
-        return Ok(splits);
+    pub async fn subscribe(self, as_of: Antichain<T>) -> Result<Subscribe<K, V, T, D>, Since<T>> {
+        trace!("ReadHandle::subscribe as_of={:?}", as_of);
+
+        let () = self.machine.verify_listen(&as_of).await?;
+        Ok(Subscribe::new(self, as_of).await)
```

Comment:
I think Subscribe::new probably wants to take the Vec of batches + Listen, so we don't have to duplicate stuff like this `verify_listen` call (so just move the `read.snapshot` and `read.listen` call that you currently have in `new` to here)